[IMP] mail: discuss sidebar compact mode

alexkuhn opened the pull request:
This commit adds a new "arrow" button in discuss sidebar panel, next to "Start a meeting". Clicking on this button enables a more compact version of discuss sidebar.

The toggling of this mode is persistent on the device, thanks to local storage.

Some parts of UI are particular in compact mode. Notably:
- some data are hidden in hover floating menu
- call participants in compact mode only shows a single participant: the most recent "active" participant. Most of the time this is the last person that has talked/is talking, but when no one talks this is a participant that shares screen.

This commit also makes a few other sidebar improvements:
- improved outline of call participant cards in white theme
- call participant in sidebar show "live" when screen-sharing
- when there's a quick search value, input/button has special style
- quick search has a clear button next to input
- more contrast between sidebar/header and thread in discuss app
- discuss sidebar shows a dot next to conversation with unread messages.

https://github.com/odoo/enterprise/pull/67607

![compact-sidebar](https://github.com/user-attachments/assets/d0636a25-1947-47c1-9a79-e9c07a39816f)

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/174985.png)](https://mergebot.odoo.com/odoo/odoo/pull/174985)

phenix-factory commented on addons/mail/static/src/discuss/call/public_web/discuss_sidebar_call_participants.js:
```diff
@@ -1,28 +1,73 @@
+        /** @type {import("models").Thread} */
+        const thread = this.props.thread;
+        const sessions = [...thread.rtcSessions];
+        if (this.compact) {
+            sessions?.sort((s1, s2) => {
+                if (s1.isActuallyTalking && !s2.isActuallyTalking) {
+                    return -1;
+                }
+                if (!s1.isActuallyTalking && s2.isActuallyTalking) {
+                    return 1;
+                }
+                if (s1.isVideoStreaming && !s2.isVideoStreaming) {
+                    return -1;
+                }
+                if (!s1.isVideoStreaming && s2.isVideoStreaming) {
+                    return 1;
+                }
+                return s2.talkingTime - s1.talkingTime;
+            });
+            if (sessions.length === 0) {
+                return [];
+            }
+            return [sessions[0]];
+        }
```
Could this be inside it's own getter `lastActifParticipant` to split the behavior properly?

alexkuhn replied:
Done: getter `lastActiveSession`, and `sessions` (previously `participants`, renamed for clarity of model)

phenix-factory commented on addons/mail/static/src/core/public_web/discuss_sidebar.js:
```diff
@@ -12,11 +16,22 @@ export const discussSidebarItemsRegistry = registry.category("mail.discuss_sideb
+
+    get compactBtnText() {
+        if (this.store.discuss.isSidebarCompact) {
+            return _t("Make panel bigger");
+        }
+        return _t("Make panel smaller");
```
Shouldn't we explicitly name this `sidebar` instead of `panel`?  For translation context?
I get that it probably already exists, but I'm not sure that reusing the translation string is a good thing in master.

alexkuhn replied:
Sidebar is a relatively good name when there was no "inspector", i.e. panel on the right.
Users do not get this component is named sidebar, this is a technical name. They see it as the left panel, and since "panel" is quite clear, I choose this name. The button is on the panel so there's no ambiguity with the right panel.

phenix-factory commented on addons/mail/static/src/core/web/discuss_sidebar_mailboxes.xml:
```diff
@@ -2,35 +2,44 @@
-            <div class="me-2 text-truncate" t-esc="mailbox.name"/>
-            <div t-attf-class="flex-grow-1 {{ mailbox.counter === 0 ? 'me-3': '' }}"/>
+            <ThreadIcon className="'bg-inherit ' + (store.discuss.isSidebarCompact ? '' : 'ms-3 me-2')" thread="mailbox"/>
+            <div t-if="!store.discuss.isSidebarCompact" class="me-2 text-truncate" t-esc="mailbox.name"/>
+            <div t-if="!store.discuss.isSidebarCompact" t-attf-class="flex-grow-1 {{ mailbox.counter === 0 ? 'me-3': '' }}"/>
```
Maybe?
```suggestion
            <t t-if="!store.discuss.isSidebarCompact">
                <div  class="me-2 text-truncate" t-esc="mailbox.name"/>
                <div t-attf-class="flex-grow-1 {{ mailbox.counter === 0 ? 'me-3': '' }}"/>
            </t>
```

phenix-factory commented on addons/mail/static/src/discuss/call/public_web/discuss_sidebar_call_participants.js:
```diff
@@ -1,28 +1,74 @@
                 ) ?? 1
             );
         });
+        return sessions;
```
Could directly return `session.sort...` instead

phenix-factory commented on addons/mail/static/src/discuss/core/public_web/discuss_sidebar_categories.js:
```diff
@@ -87,25 +110,108 @@ export class DiscussSidebarCategories extends Component {
+    setup() {
+        super.setup();
+        this.store = useState(useService("mail.store"));
+        this.discusscorePublicWebService = useState(useService("discuss.core.public.web"));
```
`discussCorePublicWebService` ?

alexkuhn replied:
Not a big deal, was already there. Besides, I would like to kill that service so it feels stupid to rename the variable in the meantime.

phenix-factory commented:
This part feels a bit messy to me:
![image](https://github.com/user-attachments/assets/48022274-957b-4585-8f74-e8ecb4567eac)

I would: 
- Reduce and intend the size of the avatar of the connected people. It will be easier to understand that this is a sub-section of the channel.
- I think the 1 is the number of people inside the call? A badge would be nice or some other form of background.
- The microphone icon should be closer to the guest avatar (like the unread message counter?)
- Do we need this "sound" icon on the left now that we have a list of the people inside the call? 
- I understand the border, but with the proposed changes, It might be unnecessary.

phenix-factory commented on addons/mail/static/src/discuss/call/public_web/discuss_sidebar_call_participants.js:
```diff
@@ -1,23 +1,68 @@
-        return this.props.thread.rtcSessions?.sort((s1, s2) => {
+    get sessions() {
+        /** @type {import("models").Thread} */
+        const thread = this.props.thread;
+        const sessions = [...thread.rtcSessions];
```
```suggestion
        const sessions = [...this.props.thread.rtcSessions];
```

alexkuhn replied:
I did this on purpose for proper typing of `thread`. This helps with making less programming errors :)

phenix-factory replied:
In this case, it would be better to directly type de props of the component

alexkuhn replied:
This has been done in this component, but somehow it doesn't work. `LinkPreviewConfirmDelete` component did the same and it works. If you know why the typing of prop doesn't work here then no problem for the change :)

alexkuhn replied:
Good catch, thanks!

alexkuhn commented:
> This part feels a bit messy to me: ![image](https://private-user-images.githubusercontent.com/1810149/355087775-48022274-957b-4585-8f74-e8ecb4567eac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4Njg3MTYsIm5iZiI6MTcyMjg2ODQxNiwicGF0aCI6Ii8xODEwMTQ5LzM1NTA4Nzc3NS00ODAyMjI3NC05NTdiLTQ1ODUtOGY3NC1lOGVjYjQ1NjdlYWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDVUMTQzMzM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTdhMzI2Zjg4MDJmZmRlNWNlZWNjOThhNGMxZTA5ZTBhZDE1N2Q4ZTgyOTdmNTU1NDNkNTAyZWFlMzY2M2E1OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.guO0hntVv6CHa0zBvsaUME_rRgrDEniief9ozuLy7d4)
> 
> I would:
> 
> * Reduce and intend the size of the avatar of the connected people. It will be easier to understand that this is a sub-section of the channel.
> * I think the 1 is the number of people inside the call? A badge would be nice or some other form of background.
> * The microphone icon should be closer to the guest avatar (like the unread message counter?)
> * Do we need this "sound" icon on the left now that we have a list of the people inside the call?
> * I understand the border, but with the proposed changes, It might be unnecessary.

I made the following changes:
- I kept border otherwise it's confusing, but I reduced the opacity.
- I removed the "in-call" counter. We can see all participants with mouse-hovering so we don't really need the counter.
- I moved participant status correctly closer to avatar
- calling icon is kept, but I moved it placement slightly and background so that it puts inherited bg only when actual icon overlaps

In the end this looks like this:
<img width="58" alt="Screenshot 2024-08-05 at 16 44 58" src="https://github.com/user-attachments/assets/19d5efc5-5828-4319-9e21-5e9c4788b92f"> <img width="61" alt="Screenshot 2024-08-05 at 16 47 45" src="https://github.com/user-attachments/assets/a2ec6225-42e9-48a4-b42b-bda9fcbbf7cb"> <img width="60" alt="Screenshot 2024-08-05 at 16 47 57" src="https://github.com/user-attachments/assets/0e3fcc81-81c7-4a9d-a32f-b60bcd67bc2f">

phenix-factory reviewed (approved):
@robodoo r+
